ibc-core/ics24-host/types/src/identifiers/chain_id.rs

```diff
@@ -116,6 +115,173 @@ impl ChainId {
     }
 }
 
+#[cfg(feature = "serde")]
+impl<'de> Deserialize<'de> for ChainId {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        const FIELDS: &[&str] = &["id", "revision_number"];
```

Comment:
Why not something like: 

```rust
#[derive(serde::Deserialize)]
struct IdAndRevision {
    id: String,
    revision_number: u64,
}

let IdAndRevision { id, revision_number } =
    IdAndRevision::deserialize(D)?;
if revision_number == parse_revision_number(chain_id.as_str())? {
    Ok(Self { id, revision_number })
} else {
    Err(...)
}
```

And add `fn parse_revision_number(id: &str) -> Result<u64, IdentifierError>;` function

Reply:
It's not really clear to me which part of the `serde::Deserialize` impl you're thinking your suggested code should replace.

Reply:
This would be body of the `deserialize` method:

```rust
#[cfg(feature = "serde")]
impl<'de> Deserialize<'de> for ChainId {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        struct IdAndRevision {
            id: String,
            #[serde(alias = "revisionnumber")]
            revision_number: u64,
        }

        let IdAndRevision { id, revision_number } =
            IdAndRevision::deserialize(D)?;
        if revision_number == parse_revision_number(chain_id.as_str())? {
            Ok(Self { id, revision_number })
        } else {
            Err(...)
        }
    }
}
```

Reply:
This would be body of the `deserialize` method:

```rust
#[cfg(feature = "serde")]
impl<'de> Deserialize<'de> for ChainId {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        struct IdAndRevision {
            id: String,
            revision_number: u64,
        }

        let IdAndRevision { id, revision_number } =
            IdAndRevision::deserialize(D)?;
        if revision_number == parse_revision_number(chain_id.as_str())? {
            Ok(Self { id, revision_number })
        } else {
            Err(...)
        }
    }
}
```

Reply:
Still `IdAndRevision::deserialize(deserializer)?` is not working properly for all cases. Anyhow let’s leave further optimization for later PRs.

Reply:
What’s the issue? IdAndRevision with a derive should have the exact same deserialisation logic as ChainId.